src/org/exist/xslt/EXistURIResolver.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * eXist Open Source Native XML Database
+ * Copyright (C) 2001-2017 The eXist Project
+ * http://exist-db.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package org.exist.xslt;
+
+import java.io.IOException;
+import java.lang.reflect.Array;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Path;
+import javax.xml.transform.Source;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.URIResolver;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamSource;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.exist.dom.persistent.BinaryDocument;
+import org.exist.dom.persistent.DocumentImpl;
+import org.exist.security.Permission;
+import org.exist.security.PermissionDeniedException;
+import org.exist.storage.BrokerPool;
+import org.exist.storage.DBBroker;
+import org.exist.xmldb.XmldbURI;
+
+/**
+ * Implementation of URIResolver which
+ * will resolve paths from the eXist database
+ */
+public class EXistURIResolver implements URIResolver {
+
+  private static final Logger LOG = LogManager.getLogger(EXistURIResolver.class);
+
+  final BrokerPool db;
+  final String basePath;
+
+  public EXistURIResolver(final BrokerPool db, final String docPath) {
+    this.db = db;
+    this.basePath = normalize(docPath);
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("EXistURIResolver base path set to " + basePath);
+    }
+  }
+
+  private String normalize(String uri) {
+    if (uri.startsWith(XmldbURI.EMBEDDED_SERVER_URI_PREFIX)) {
+      return uri.substring(XmldbURI.EMBEDDED_SERVER_URI_PREFIX.length());
+    }
+
+    return uri;
+  }
+
+  /**
+   * Simplify a path removing any "." and ".." path elements.
+   * Assumes an absolute path is given.
+   */
+  private String normalizePath(final String path) {
+    if (!path.startsWith("/")) {
+      throw new IllegalArgumentException("normalizePath may only be applied to an absolute path; " +
+          "argument was: " + path + "; base: " + basePath);
+    }
+
+    final String[] pathComponents = path.substring(1).split("/");
```

Comment:
This will cause an NPE:
```java
final String path = "/";

...

if (!path.startsWith("/")) {
  final String[] pathComponents = path.substring(1).split("/");
}
```